### DISCUSSION TITLE
Creating huge xarray datasets

### DISCUSSION THREAD

**Discussion by muaali** [pydata/xarray#5367] [Q&A]
Hi.

I have a lot of satellite images in my local directory. I want to create an xarray dataset from `n` satellite images (each satellite image is huge in memory). Normally, one would like to iteratively create xarray data array objects, concatenate them and then convert to xarray dataset. But this approach leads to memory issues as `xr.concat` forces dask arrays to load (thread: https://github.com/pydata/xarray/issues/4628) 

What would be the right approach to solve this issue with xarrays and dask ?

Few things to note:
- Images are of extension `.tif`
- Images are not of the same size
- Each image can be up to 2 GB and I would like to scale to as many images as possible
- using `xr.open_rasterio` leads to dataset with max image (image with the maximum size) repeated `n` times where `n` is the number of images. I will have to resize each image to save size and then save back, which is a tedious work and does not seem an efficient way to accomplish this task.

Below is my code on high level, but it I still run into memory issues with even after chunking with `dask`

```
xdarrays_list = []


def create_xdarray_with_raster(fpath):
    rasterio_container = rasterio.open(fpath)
    Iarray = np.squeeze(rasterio_container.read()) 
    xdarray = xr.DataArray(Iarray, coords=[np.arange(Iarray.shape[0]), np.arange(Iarray.shape[1])], dims=["x", "y"])
    xdarray.attrs = rasterio_container.tags()

    return xdarray


for fpath in fpaths:
    raster_xdarray = create_xdarray_with_raster(raster_fpath).chunk(4096,4096)
    assert(dask.is_dask_collection(raster_xdarray))
    xdarrays_list.append(raster_xdarray)

xr.Dataset({"values": xr.concat(xdarrays_list, dim="band", data_vars="all", fill_value=-1)},
            coords={
                "latitude": list_of_lats, #of len same as number of images
                "longitude": list_of_longs, #of len same as number of images
            },
        )
```

**Comment by keewis** [comment]
that thread is about lazy concatenation *without* using `dask`, with `dask` it should work just fine:
```python
In [10]: import xarray as xr
    ...: from xarray.tests import raise_if_dask_computes
    ...: 
    ...: ds = xr.tutorial.open_dataset("rasm").chunk({"time": 8, "x": 5, "y": 5})
    ...: 
    ...: with raise_if_dask_computes(max_computes=0):
    ...:     ds1 = ds.isel(time=slice(0, 18))
    ...:     ds2 = ds.isel(time=slice(18, None))
    ...: 
    ...:     concat = xr.concat([ds1, ds2], dim="time")
    ...: 
    ...: xr.testing.assert_identical(ds, concat)

In [11]:
```

Instead of doing this manually using a loop you should also be able to use `open_mfdataset` (if you need to use `rasterio` to load your data this requires `xarray>=0.18` and a recent version of `rioxarray`).

**Reply by muaali** [reply]
@keewis I get this error: 
`RuntimeError: Too many computes. Total: 1 > max: 0.`

I did use inbuilt support of `rasterio` in xarray  but it was giving me float indices so I wrote something small myself. 
So this is how my code looks on a high level:

```
xdarrays_list = []


def create_xdarray_with_raster(fpath, load_metadata=True):
    rasterio_container = rasterio.open(fpath)
    Iarray = np.squeeze(rasterio_container.read()) 
    xdarray = xr.DataArray(Iarray, coords=[np.arange(Iarray.shape[0]), np.arange(Iarray.shape[1])], dims=["x", "y"])
    xdarray.attrs = rasterio_container.tags()

    return xdarray


for fpath in fpaths:
    raster_xdarray = create_xdarray_with_raster(raster_fpath, load_metadata=True).chunk(4096,4096)
    assert(dask.is_dask_collection(raster_xdarray))
    xdarrays_list.append(raster_xdarray)

xr.Dataset({"intensity": xr.concat(xdarrays_list, dim="band", data_vars="all", fill_value=-1)},
            coords={
                "latitude": list_of_lats,
                "longitude": list_of_longs,
            },
        )

```

**Reply by keewis** [reply]
> I get this error: `RuntimeError: Too many computes. Total: 1 > max: 0.`

with what code? The one I posted or something else?

> but it was giving me float indices so I wrote something small myself

If I understand the documentation of `rasterio.DatasetReader.read` correctly, that's something you want to avoid calling immediately because it loads all the data into memory (and not just the indexes / metadata).

As for the `open_mfdataset` call, you *can* add a preprocessor to create integer indexes:
```python
def integer_indexes(ds):
    return (
        ds.assign_coords(xi=np.arange(ds.sizes["x"]), yi=np.arange(ds.sizes["y"]))
        .swap_dims({"x": "xi", "y": "yi"})
    )

ds = xr.open_mfdataset(fpaths, engine="rasterio", preprocess=integer_indexes)
```
but note that the float indexes are most likely part of your data so you may actually want to keep that (see for example the output of `xr.tutorial.open_rasterio("RGB.byte")`). Also, integer indexes are usually not necessary because you can use `isel` instead of `sel` for position-based indexing and for alignment you have to be sure the data is always on the same grid.

**Reply by muaali** [reply]
@keewis 


> with what code? The one I posted or something else?

yes with the one that you posted above.

With `open_mfdataset` I get the following error:
```
ValueError: unrecognized engine for open_dataset: rasterio
must be one of: ['netcdf4', 'scipy', 'pydap', 'h5netcdf', 'pynio', 'pseudonetcdf', 'cfgrib', 'zarr']
``` 
The reason being that I am working on `.tif` file formats. One alternative I was thinking was to convert each `.tif` image into `xr.dataset` and save as `.nc`. Then once I have saved `xr.dataset` all of images as `.nc` files, I could load all of them from a directory using `open_mfdataset` using chunks and then save back as one huge `.nc`. That might work but seems very inefficient way to do it.



The last method I tried is to use inbuilt `xarray.open_rasterio()` to open `.tif` files without explicity calling `read` method but still my kernel dies because of memory issues.
```
# tif_fpaths is list of paths of tif files.
list_of_data_arrays = []

for tif_fpath in tif_fpaths:
    da = xr.open_rasterio(tif_fpath, chunks=(1024,1024))
    list_of_data_arrays.append(da)

ds = xr.Dataset({"intensity": xr.concat(list_of_data_arrays, dim="band", data_vars="all", fill_value=-1)},
            coords={
                "latitude": list_of_lats, #from metadata
                "longitude": list_of_longs, #from metadata
            },
        )
ds.to_netcdf('huge_cube.nc') # <- This is where kernel dies.
```

**Reply by keewis** [reply]
> yes with the one that you posted above.

that might be a bug. Could you share the versions of `xarray` and `dask` you have installed?

Edit: if possible, the output of `xr.show_versions()` would be even better

> With `open_mfdataset` I get the following error:

I don't know too much about memory consumption while writing to disk, but the reason why `open_mfdataset` does not work is because you don't have `rioxarray` installed. To be clear, you need `xarray>=0.18` and `rioxarray>=0.4` for that to work.

**Reply by muaali** [reply]
INSTALLED VERSIONS
------------------
commit: None
python: 3.7.10 (default, Feb 26 2021, 18:47:35) 
[GCC 7.3.0]
python-bits: 64
OS: Linux
OS-release: 5.4.0-64-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8
libhdf5: 1.10.6
libnetcdf: 4.7.4

xarray: 0.17.0
pandas: 1.2.4
numpy: 1.20.2
scipy: 1.6.3
netCDF4: 1.5.6
pydap: None
h5netcdf: None
h5py: 3.2.1
Nio: None
zarr: 2.8.1
cftime: 1.4.1
nc_time_axis: None
PseudoNetCDF: None
rasterio: 1.2.2
cfgrib: None
iris: None
bottleneck: 1.3.2
dask: 2021.04.1
distributed: 2021.04.1
matplotlib: 3.4.1
cartopy: None
seaborn: None
numbagg: None
pint: None
setuptools: 52.0.0.post20210125
pip: 21.0.1
conda: 4.10.1
pytest: 6.2.4
IPython: 7.22.0
sphinx: None

**Reply by max-sixty** [reply]
> One alternative I was thinking was to convert each `.tif` image into `xr.dataset` and save as `.nc`. Then once I have saved `xr.dataset` all of images as `.nc` files, I could load all of them from a directory using `open_mfdataset` using chunks and then save back as one huge `.nc`. That might work but seems very inefficient way to do it.

I have done something similar to this, for files that xarray doesn't natively load. It worked reasonably.

**Reply by dcherian** [reply]
I believe rioxarray is the solution here: https://corteva.github.io/rioxarray/stable/index.html I think they have a `mfdataset` version that can read tif files. With newer xarray and rioxarray, I think you could just pass `engine="rioxarray"` to `open_dataset`

cc @snowman2

**Reply by snowman2** [reply]
`engine="rasterio"`: https://corteva.github.io/rioxarray/stable/getting_started/getting_started.html

**Reply by muaali** [reply]
Alright. so with `xarray` updated to `0.18.2` and `rioxarray` installed. This line works : 
```
ds1 = xr.open_mfdataset(tif_fpaths, engine="rasterio")
```
But facing this issue where I only get one 1 band in `ds1` where I will expect number of bands to be: `num_bands=len(tif_fpaths)`
Same issue takes place when I try to `open_mfdataset` all saved `netcdf4` files from a directory `ds2 = xr.open_mfdataset(nc_fpaths, combine='by_coords', engine='netcdf4')`. It only returns me `ds2` with one band.

Any idea why is that so ?

**Reply by muaali** [reply]
@max-sixty I would be happy to have take a look at that code. Might be helpful.

**Reply by snowman2** [reply]
> But facing this issue where I only get one 1 band in ds1 where I will expect number of bands to be: num_bands=len(tif_fpaths)

Is there a new dimension/coordinate added?

**Reply by muaali** [reply]
@snowman2 All of images don't have the same size so that leads to `xr.DataArray` of different sizes. Could that cause this issue ?

**Reply by snowman2** [reply]
@muaali you will want to ensure they are all the same projection/resolution if you want to stack them. This can help you prepare them: https://corteva.github.io/rioxarray/stable/examples/reproject_match.html

**Comment by muaali** [comment]
@keewis so I am able to create a 21 GB of `.nc` with the following code. Here `xr.concat` does not force dask arrays to load.

```
xarray_list = []
for tif_fpath in tif_fpaths:
    da = xr.open_rasterio(tif_fpath)
    da = da.chunk(4096,4096) # arbitrary chunk size
    assert(dask.is_dask_collection(da))
    xarray_list.append(da)

ds = xr.Dataset(
            {"intensity": xr.concat(xarray_list, dim="band", data_vars="all", fill_value=-1)},
            coords={
                "latitude": list_of_lats, #derived from metadata
                "longitude": list_of_longs, #derived from metadata
            },
        )
ds.to_netcdf('huge_cube.nc'))
```

But what I don't like about this solution is I am heavily dependent on default functionalities of `xr.open_rasterio`. So if I would like to perform some preprocessing or create another dataset variable (apart from `intensity`) based on some external information, I would be back to square one. So I am just thinking how to modify following function to achieve that (and I could replicate this for others):
```
def create_xdarray_with_raster(fpath, load_metadata=True):
    rasterio_container = rasterio.open(fpath)
    Iarray = np.squeeze(rasterio_container.read())  # <- This right here is the issue as you pointed out.
    xdarray = xr.DataArray(Iarray, coords=[np.arange(Iarray.shape[0]), np.arange(Iarray.shape[1])], dims=["x", "y"])
    xdarray.attrs = rasterio_container.tags()

    return xdarray
```

**Reply by muaali** [reply]
@snowman2  since you are live I am gonna tag you here as well :)

**Reply by snowman2** [reply]
@muaali have you looked into using [rioxarray.open_rasterio](https://corteva.github.io/rioxarray/stable/rioxarray.html#rioxarray.open_rasterio)?

**Reply by muaali** [reply]
@snowman2 yes and it quite similar to what is implemented in `xarray.open_rasterio`. All I am looking for at the moment, is how to perform dask lazy operations without running into memory issues with file formats (or just loaded arrays) that are not by default supported by `rasterio`

**Reply by snowman2** [reply]
Here is an example: https://corteva.github.io/rioxarray/stable/examples/dask_read_write.html

**Reply by muaali** [reply]
Can `xr.open_mfdataset` be used to stack `nc` datasets along variable dimension ? My problem statement is to stack `n` huge images in a cube along dataset variable. So with `n` images I end up with `n*x*y` dimensions where `n` indicates number of images and `x`,`y` are dimensions for each image.

**Reply by kmuehlbauer** [reply]
@muaali If you are interested only in the `.nc` output file you might use `nco` to concatenate the single files into one.

**Reply by muaali** [reply]
What is `nco` ?

I am unable to find a way to stack images along `dataset.variable` (even with using `dask` not sure if I am doing something wrong or it's not supported. I run into memory issues). Saving each image's `xr.dataset` to `.nc` and then opening with `open_mfdataset` seemed like a way to go but now I realize this function applies concatenation only along `dims` and I cannot stack `xr.dataArrays`

**Reply by kmuehlbauer** [reply]
NCO - NetCDF Operators https://github.com/nco/nco

I've used `ncrcat` in the past, but I'm not sure if this fit's to your data. Ping @czender, who knows way more about `nco`.

**Reply by kmuehlbauer** [reply]
For your `open_mfdataset`-approach you might need to use `preprocess`-kwarg and add the needed concatenation dimension and coordinate to your source data (kind of what @keewis suggested above).

**Reply by muaali** [reply]
But then again, `open_mfdataset` does not concatenate each dataset along a new dimension or does it ? I tried setting keyword argument `concat_dim=[..., None, ...]` but ran into some errors.

with combine='nested':
`ValueError: concat_dims has length 3 but the datasets passed are nested in a 1-dimensional structure`
with default `combine="by_coords"`:
`ValueError: Could not find any dimension coordinates to use to order the datasets for concatenation`

**Reply by muaali** [reply]
Each dataset loaded from `.nc` looks like this. They all have same dimension sizes and same coord sizes:
![image](https://user-images.githubusercontent.com/69156640/119631112-40c3d300-be18-11eb-89d3-f5734c6072f5.png)

**Reply by kmuehlbauer** [reply]
@muaali In general `open_mfdataset` will concat along a new dimension when specified if the source datasets can be concatenated (other dimensions are in agreement). If that's not the case you would need the approach suggested by @keewis above. Are the x,y coordinates the same on all datasets? Is the grid the same?

**Reply by kmuehlbauer** [reply]
OK, your message above already answers my question.

**Reply by kmuehlbauer** [reply]
So, doesn't `ds = xr.open_mfdataset(nc-filelist, concat_dim="new_dim", chunks={"x": x_size, "y": y_size})` work for you?

You would need to set some chunk-sizes which works best, If you omit the `chunks`-kwarg then the size would be (26646, 17762) as in your dataset (full-data).

**Reply by muaali** [reply]
I get this error:
`ValueError: Could not find any dimension coordinates to use to order the datasets for concatenation`

Just to clarify, `nc_filelist = ["path/to/nc1.nc","path/to/nc2.nc",....,]` is a list of path/to/nc files.

**Reply by muaali** [reply]
But this seems to work apparently: 
`all_ds = xr.open_mfdataset(nc_fpaths, concat_dim="new_dim", combine='nested', engine='netcdf4')`

**Reply by kmuehlbauer** [reply]
OK, can you show `print(all_ds)`?

**Reply by muaali** [reply]
sure. 
![image](https://user-images.githubusercontent.com/69156640/119637218-1b39c800-be1e-11eb-8bcd-eb5f501669a3.png)

**Reply by kmuehlbauer** [reply]
Looks pretty nice, exactly what it should. Have you already tried to export it?

**Reply by muaali** [reply]
Now when I do this:
`all_ds.to_netcdf(os.path.join(save_dir, 'huge_cube.nc'))`

I run into memory issues again even after chunking. My CPU memory slowly keeps increasing until it is killed. Arghh

**Reply by kmuehlbauer** [reply]
I'm not really sure I understand you OP correctly. But looking at the lat/lon coordinates, the grids are not the same. You would have to keep that in mind when working with the data (selecting data etc).

**Reply by kmuehlbauer** [reply]
> Now when I do this:
> `all_ds.to_netcdf(os.path.join(save_dir, 'huge_cube.nc'))`
> 
> I run into memory issues again even after chunking. My CPU memory slowly keeps increasing until it is killed. Arghh

Did you specify chunk-sizes in the call to `open_mfdataset`? If so, which?

**Reply by muaali** [reply]
> I'm not really sure I understand you OP correctly. But looking at the lat/lon coordinates, the grids are not the same. You would have to keep that in mind when working with the data (selecting data etc).

So the way I have approached it is that for each image, there is `coordinate.latitude` , `coordinate.longitude`, `coordinate.incidence_angle`. The idea is to perform slicing along 3rd dimension for images that are at certain locations and within certain angles. Do you think it is a wrong approach. What could be an alternative ?

**Reply by muaali** [reply]
> Did you specify chunk-sizes in the call to open_mfdataset? If so, which?

`all_ds = xr.open_mfdataset(nc_fpaths, concat_dim="new_dim", combine='nested', chunks={"x": 4096, "y": 4096}, engine='netcdf4')`

**Reply by kmuehlbauer** [reply]
> all_ds = xr.open_mfdataset(nc_fpaths, concat_dim="new_dim", combine='nested', chunks={"x": 4096, "y": 4096}, engine='netcdf4')

Just for the sake of testing, can you drop/delete the coordinates (lat/lon/incident_angle) before exporting to netcdf?

**Reply by kmuehlbauer** [reply]
> > I'm not really sure I understand you OP correctly. But looking at the lat/lon coordinates, the grids are not the same. You would have to keep that in mind when working with the data (selecting data etc).
> 
> So the way I have approached it is that for each image, there is `coordinate.latitude` , `coordinate.longitude`, `coordinate.incidence_angle`. The idea is to perform slicing along 3rd dimension for images that are at certain locations and within certain angles. Do you think it is a wrong approach. What could be an alternative ?

Unfortunately my data sources are fixed and don't move around that much (ground weather radar). But some general notes: You might be able to slice along the third dimension, but still you have to keep in mind, that your x,y coordinates are only identical if all three of your coordinates are the same (for the different slices). Also possible, that slicing isn't possible at all, since the data is not properly ordered. I'm sorry, but this as far as I can help here.

**Reply by muaali** [reply]
Thanks alot @kmuehlbauer for your help so far. I tried deleting lats/long/angles but still memory chokes after a while.
